src/modules/fancyzones/editor/FancyZonesEditor/CanvasZone.xaml.cs

```diff
@@ -16,192 +17,336 @@ namespace FancyZonesEditor
+        public CanvasZone()
+        {
+            InitializeComponent();
+            Canvas.SetZIndex(this, zIndex++);
+        }
 
-        private readonly Settings _settings = ((App)Application.Current).ZoneSettings;
+        private CanvasLayoutModel model;
 
-        private static readonly int _minZoneWidth = 64;
-        private static readonly int _minZoneHeight = 72;
-        private static int _zIndex = 0;
+        private int zoneIndex;
 
-        public CanvasZone()
+        public enum ResizeMode
         {
-            InitializeComponent();
-            Panel.SetZIndex(this, _zIndex++);
+            ModeLow = 1,
+            ModeHigh = 2,
+            ModeBoth = 3,
         }
 
-        private void Move(double xDelta, double yDelta)
+        private abstract class SnappyHelperBase
         {
-            Int32Rect rect = Model.Zones[ZoneIndex];
-            if (xDelta < 0)
+            public int ScreenW { get; private set; }
+
+            protected List<int> Snaps { get; private set; }
+
+            protected int MinValue { get; private set; }
+
+            protected int MaxValue { get; private set; }
+
+            public int Position { get; protected set; }
+
+            public ResizeMode Mode { get; private set; }
+
+            /*
+            public const int ModeLow = 1;
+            public const int ModeHigh = 2;
+            public const int ModeBoth = 3;
+            */
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="SnappyHelperBase"/> class.
+            ///     Just pass it the canvas arguments. Use mode
+            ///     to tell it which edges of the existing masks to use when building its list
+            ///     of snap points, and generally which edges to track. There will be two
+            ///     SnappyHelpers, one for X-coordinates and one for
+            ///     Y-coordinates, they work independently but share the same logic.
+            /// </summary>
+            /// <param name="zones">The list of rectangles describing all zones</param>
+            /// <param name="zoneIndex">The index of the zone to track</param>
+            /// <param name="isX"> Whether this is the X or Y SnappyHelper</param>
+            /// <param name="mode"> One of the three modes of operation (for example: tracking left/right/both edges)</param>
+            /// <param name="screen_w"> The size of the screen in this (X or Y) dimension</param>
+            public SnappyHelperBase(IList<Int32Rect> zones, int zoneIndex, bool isX, ResizeMode mode, int screen_w)
             {
-                xDelta = Math.Max(xDelta, -rect.X);
-            }
-            else if (xDelta > 0)
-            {
-                xDelta = Math.Min(xDelta, _settings.WorkArea.Width - rect.Width - rect.X);
-            }
+                int track_p = isX ? zones[zoneIndex].X : zones[zoneIndex].Y;
+                int track_w = isX ? zones[zoneIndex].Width : zones[zoneIndex].Height;
```

Comment:
`track_w` is `zone_coordinate_width`

Reply:
I don't want to imply a particular direction, changing this to zone_axis_size

Reply:
OK!